File: OpenInCodeCore.m
#import "OpenInCodeCore.h"

NSString * const OICVSCodeBundleIdentifier = @"com.microsoft.VSCode";
NSString * const OICVSCodeInsidersBundleIdentifier = @"com.microsoft.VSCodeInsiders";

NSArray<NSString *> *OICPreferredVSCodeBundleIdentifiers(void)
{
    return @[OICVSCodeBundleIdentifier, OICVSCodeInsidersBundleIdentifier];
}

NSString *OICPathForFinderURL(NSURL *url)
{
    if (url == nil || ![url isFileURL]) {
        return nil;
    }

    NSData *bookmark = [NSURL bookmarkDataWithContentsOfURL:url error:nil];
    if (bookmark != nil) {
        NSURL *resolvedURL = [NSURL URLByResolvingBookmarkData:bookmark
                                                       options:NSURLBookmarkResolutionWithoutUI
                                                 relativeToURL:nil
                                           bookmarkDataIsStale:nil
                                                         error:nil];
        if (resolvedURL != nil) {
            url = resolvedURL;
        }
    }

    NSString *path = [[url path] stringByExpandingTildeInPath];
    if ([path length] == 0) {
        return nil;
    }

    BOOL isDirectory = NO;
    if (![[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&isDirectory]) {
        return nil;
    }

    NSURL *packageCheckURL = [url URLByResolvingSymlinksInPath];
    NSNumber *isPackageValue = nil;
    [packageCheckURL getResourceValue:&isPackageValue forKey:NSURLIsPackageKey error:nil];
    BOOL isPackage = [isPackageValue boolValue];

    if (!isDirectory || isPackage) {
        path = [path stringByDeletingLastPathComponent];
    }

    return [path length] > 0 ? path : nil;
}

File: Sources/OpenInCode/OpenInCodeCore.swift
import Foundation

let OICVSCodeBundleIdentifier = "com.microsoft.VSCode"
let OICVSCodeInsidersBundleIdentifier = "com.microsoft.VSCodeInsiders"

func OICPreferredVSCodeBundleIdentifiers() -> [String] {
    [OICVSCodeBundleIdentifier, OICVSCodeInsidersBundleIdentifier]
}

func OICFileSystemPathForURL(_ url: URL) -> String {
    var path: String
    if #available(macOS 13.0, *) {
        path = url.path(percentEncoded: false)
    } else {
        path = url.path
    }

    while path.count > 1 && path.last == "/" {
        path.removeLast()
    }
    return path
}

func OICPathForFinderURL(_ finderURL: URL?) -> String? {
    guard var url = finderURL, url.isFileURL else {
        return nil
    }

    let resourceValues = try? url.resourceValues(forKeys: [.isAliasFileKey, .isSymbolicLinkKey])
    if resourceValues?.isAliasFile == true && resourceValues?.isSymbolicLink != true {
        guard let bookmarkData = try? URL.bookmarkData(withContentsOf: url) else {
            return nil
        }

        var isStale = false
        guard let resolvedURL = try? URL(
            resolvingBookmarkData: bookmarkData,
            options: .withoutUI,
            relativeTo: nil,
            bookmarkDataIsStale: &isStale
        ) else {
            return nil
        }
        url = resolvedURL
    }

    var path = (OICFileSystemPathForURL(url) as NSString).expandingTildeInPath
    guard !path.isEmpty else {
        return nil
    }

    var isDirectory: ObjCBool = false
    guard FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory) else {
        return nil
    }

    let packageCheckURL = url.resolvingSymlinksInPath()
    let packageValues = try? packageCheckURL.resourceValues(forKeys: [.isPackageKey])
    let isPackage = packageValues?.isPackage == true

    if !isDirectory.boolValue || isPackage {
        path = (path as NSString).deletingLastPathComponent
    }

    return path.isEmpty ? nil : path
}
